tests/shouldwork/Cores/Xilinx/Floating/TH.hs

```diff
@@ -5,58 +5,94 @@ License    :  BSD2 (see the file LICENSE)
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE ViewPatterns #-}
 
 module Floating.TH where
 
-import Clash.Prelude (natToNum, unpack, Signed, Unsigned)
+import Clash.Prelude
+  ( Signed, Unsigned, BitVector, (++#), ShowX(..), BitPack
+  , msb, testBit, truncateB, showsPrecXWith, natToNum, unpack, pack
+  )
 
 import Prelude
+import Data.Coerce (coerce)
+import Numeric (showHex)
 import Numeric.IEEE
-  (epsilon, infinity, maxFinite, minDenormal, minNormal)
+  (epsilon, infinity, maxFinite, minDenormal, minNormal, nan)
 
 import Clash.Cores.Xilinx.Floating as F
 import Clash.Cores.Xilinx.Floating.Internal as F
 
+newtype FloatVerifier = FloatVerifier Float
+  deriving newtype (BitPack)
+
+instance Eq FloatVerifier where
+  (FloatVerifier x) == (FloatVerifier y) = pack x == pack y
+
+instance ShowX FloatVerifier where
+  showsPrecX = showsPrecXWith showsPrec
+
+instance Show FloatVerifier where
+  showsPrec = floatVerifierShowsPrec
+
+floatVerifierShowsPrec
+  :: Int
+  -> FloatVerifier
+  -> ShowS
+floatVerifierShowsPrec _ (FloatVerifier x)
+  | isNaN x = nanSign . nanString . showHex payload . (')':)
+  | otherwise = shows x
+ where
+  nanSign | msb (pack x) == 0 = ('+':)
+          | otherwise         = ('-':)
+  nanString
+    | testBit (pack x) 22 = ("qNaN(0x" ++)
+    | otherwise           = ("sNaN(0x" ++)
+  payload = truncateB $ pack x :: BitVector 22
+
 delayOutput
   :: Int
-  -> [(Float, Float, Float)]
-  -> [(Float, Float, Float)]
-delayOutput d es = zip3 xs ys rs
+  -> a
+  -> [(Float, Float, a)]
+  -> [(Float, Float, a)]
+delayOutput d aDef es = zip3 xs ys rs
  where
   (xs0, ys0, rs0) = unzip3 es
   xs = xs0 ++ repeat (last xs0)
   ys = ys0 ++ repeat (last ys0)
-  rs = replicate d 0 ++ rs0
+  rs = replicate d aDef ++ rs0
 
-addBasicSamples :: [(Float, Float, Float)]
+addBasicSamples :: [(Float, Float, FloatVerifier)]
 addBasicSamples =
-  delayOutput (natToNum @F.AddDefDelay) $
+  coerce $ delayOutput (natToNum @F.AddDefDelay) 0.0 $
   [ (1, 4, 5)
   , (2, 5, 7)
   , (3, 6, 9)
   ]
-  ++ addSubBasicSamples
+  ++ coerce addSubBasicSamples
   ++ cartesianProductTest model interesting interesting
   ++ nanTest
  where
   model (conditionFloat -> x) (conditionFloat -> y) = conditionFloat $ x + y
 
-subBasicSamples :: [(Float, Float, Float)]
+subBasicSamples :: [(Float, Float, FloatVerifier)]
 subBasicSamples =
-  delayOutput (natToNum @F.SubDefDelay) $
+  coerce $ delayOutput (natToNum @F.SubDefDelay) 0.0 $
   [ (1, 6, -5)
   , (2, 5, -3)
   , (3, 4, -1)
   ]
-  ++ map (\(a,b,c) -> (a, negate b, c)) addSubBasicSamples
+  ++ map (\(a,b,c) -> (a, negate b, c)) (coerce addSubBasicSamples)
   ++ cartesianProductTest model interesting interesting
   ++ nanTest
  where
   model (conditionFloat -> x) (conditionFloat -> y) = conditionFloat $ x - y
 
-addSubBasicSamples :: [(Float, Float, Float)]
-addSubBasicSamples =
+addSubBasicSamples :: [(Float, Float, FloatVerifier)]
```

Comment:
You're just coercing it back and forth, this can stay `[(Float, Float, Float)]`

Reply:
Yep!